app/projectsmodel.cpp

```diff
@@ -14,192 +8,656 @@
+  if ( page == 1 )
+  {
+    // if we are populating first page, reset model and throw away previous projects
+    beginResetModel();
+    mergeProjects( merginProjects, pendingProjects, false );
+    printProjects();
+    endResetModel();
+  }
+  else
+  {
+    // paginating next page, keep previous projects and emit model add items
+    beginInsertRows( QModelIndex(), mProjects.size(), mProjects.size() + merginProjects.size() - 1 );
+    mergeProjects( merginProjects, pendingProjects, true );
+    printProjects();
+    endInsertRows();
+  }
+
+  mServerProjectsCount = projectsCount;
+  mPaginatedPage = page;
+  emit hasMoreProjectsChanged();
+}
+
+void ProjectsModel::onListProjectsByNameFinished( const MerginProjectsList &merginProjects, Transactions pendingProjects, QString requestId )
+{
+  qDebug() << "PMR: onListProjectsByNameFinished(): received response with requestId = " << requestId;
+  if ( mLastRequestId != requestId )
+  {
+    qDebug() << "PMR: onListProjectsByNameFinished(): ignoring request with id " << requestId;
+    return;
+  }
 
   beginResetModel();
-  mProjectFiles.removeAt( row );
+  mergeProjects( merginProjects, pendingProjects );
+  printProjects();
   endResetModel();
 }
 
-int ProjectModel::rowCount( const QModelIndex &parent ) const
+void ProjectsModel::mergeProjects( const MerginProjectsList &merginProjects, Transactions pendingProjects, bool keepPrevious )
 {
-  Q_UNUSED( parent );
-  return mProjectFiles.count();
+  LocalProjectsList localProjects = mLocalProjectsManager->projects();
+
+  qDebug() << "PMR: mergeProjects(): # of local projects = " << localProjects.size() << " # of mergin projects = " << merginProjects.size();
+
+  if ( !keepPrevious )
+    mProjects.clear();
+
+  if ( mModelType == ProjectModelTypes::LocalProjectsModel )
+  {
+    // Keep all local projects and ignore all not downloaded remote projects
+    for ( const auto &localProject : localProjects )
+    {
+      std::shared_ptr<Project> project = std::shared_ptr<Project>( new Project() );
+      project->local = std::unique_ptr<LocalProject>( new LocalProject( localProject ) );
+
+      MerginProject remoteEntry;
+      remoteEntry.projectName = project->local->projectName;
+      remoteEntry.projectNamespace = project->local->projectNamespace;
+
+      if ( merginProjects.contains( remoteEntry ) )
+      {
+        int i = merginProjects.indexOf( remoteEntry );
+        project->mergin = std::unique_ptr<MerginProject>( new MerginProject( merginProjects[i] ) );
+
+        if ( pendingProjects.contains( project->mergin->id() ) )
+        {
+          TransactionStatus projectTransaction = pendingProjects.value( project->mergin->id() );
+          project->mergin->progress = projectTransaction.transferedSize / projectTransaction.totalSize;
+          project->mergin->pending = true;
+        }
+        project->mergin->status = ProjectStatus::projectStatus( project );
+      }
+      else if ( project->local->localVersion > -1 )
```

Comment:
As we discuss, this may introduce confusion about remote projects - if they are from different mergin server or they are orphaned. Maybe would be good to set some specific project status for such projects - so a user can see, that they are Mergin projects, but actually cannot sync as usual with current auth session.

Reply:
I think that we actually do not need it anymore thanks to this https://github.com/lutraconsulting/input/blob/pmr-project-sync/app/projectsmodel.cpp#L623

I will double check that